### DISCUSSION TITLE
Failing Snowflake Delete Queries on Partitions

### DISCUSSION THREAD

**Discussion by kaydanzie** [dagster-io/dagster#18415] [Q&A]
Hi team! We're running into an interesting issue related to Snowflake and time-partitioned assets and we're curious if anyone else has experienced into this.

We’ve got an hourly partitioned asset that uses the `SnowflakePandasIOManager`. Our Snowflake instance allows 20 waiters to be queued for the lock on a given table. The DELETE query locks the table while it's executing. More than 20 queries waiting for a lock on a table to be released get immediately cancelled by Snowflake. Whenever we materialize an hourly partition for our asset, we expect the data in the Snowflake table for that asset to be deleted first, then the asset is materialized. This usually works as expected, except when we’re running a big backfill where more than 20 jobs may be running at once. We noticed that even if the run fails to delete the data in the partition (ie. Snowflake cancels the DELETE query because more than 20 are already queued waiting for the same table), the asset continues to materialized successfully. So we end up with duplicated data for the hourly partitions where the DELETE query failed while the rest of the asset materialization code continued to run.

Has anyone else encountered this issue? Is there a way to configure the SnowflakePandasIoManager or the asset itself to fail if it can’t delete the partition data first? Appreciate your time!

**Comment by clairelin135** [comment]
@jamiedemaria would you mind taking a look at this issue?

**Comment by jamiedemaria** [comment]
I think this is because we catch `ProgrammingError` when we run the delete clause. This is to safeguard against the table not existing, but i bet it's also catching the canceled queries and ignoring them. I think we can fix by being more specific about the error message and only ignore when it's specifically about the table not existing.

**Reply by jamiedemaria** [reply]
I can try to take a look at this next week

**Reply by kaydanzie** [reply]
Thank you Jamie! Thanks to this comment I did find the area of the code you're referring to and that does seem like an issue. We have a workaround by limiting the number of max concurrent jobs to 20 but definitely seems like it'd be an overall improvement for this to be fixed if you have the bandwidth at some point 🙏 Appreciate your responses!
